oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+        startOptimization(RandomNumberOptimizer())
+        StartOptimizationResponseDTO.newBuilder().setMessage("Started").setStarted(true).build()
+    }
+
+    sealed class SimResult(open val name: String, open val result: Map<String, Double>) {
+        data class Success(override val name: String, override val result: Map<String, Double>) : SimResult(name, result)
+        data class TimeOut(override val name: String) : SimResult(name, emptyMap())
+        data class TimeOutFailure(override val name: String, val exception: String) : SimResult(name, emptyMap())
+        data class Failure(override val name: String, val exception: String) : SimResult(name, emptyMap())
+    }
+
+    interface ValueFactory {
+        fun getInputs(inputs: List<Input>): Map<String, Double>
+    }
+
+    class RandomNumberOptimizer : ValueFactory {
+        override fun getInputs(inputs: List<Input>): Map<String, Double> = inputs.associate {
+            it.name to Random.nextDouble(it.lowerBound, it.upperBound)
+        }
+    }
+
+    /**
+     * Couple things: we are missing the overall pool of configuration, so far, everything synced are base on one node
+     * in OASIS, we have our problem definition and everything is mapped to the proxy
+     * for the case of tool1 have x1, x2 as input and f3 as output
+     * tool2 have x2, x3 f3 as input, and f4 as output will be impossible to setup in the current way though sync
+     *
+     * However, as a test bed, each evaluation request is tool dependent. You can have different configuration.
+     *
+     * The question here is when how much information is needed to a overall CFE setup relative to a evaluation. It that a
+     * bigger set of information that is need that each individual evaluation. or oasis don't need to know that at all
+     *
+     */
+    suspend fun startOptimization(valueFactory: ValueFactory) {
+        state = State.Running
+
+        for (simName in simulationsByName.keys) {
+            syncConfigFor(simName)
+        }
+
+        viewData.updateStatusMessage("Evaluating...")
+
+        while (state == State.Running) {
+            var i = 1
+            for ((simName, sim) in simulationsByName) {
+                viewData.updateStatusMessage("Evaluating: ${simName} (${i}/${simulationsByName.keys.size})")
+                val inputVector = valueFactory.getInputs(sim.inputs)
+                val message = OASISQueryDTO.newBuilder()
+                        .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
+                                .setName(simName)
+                                .putAllInputVector(inputVector)
+                        )
+                        .build()
+
+                sim.input.onNext(message)
+                val result = select<SimResult> {
+                    sim.output.onReceive { it -> SimResult.Success(it.name, it.outputVectorMap) }
+                    sim.error.onReceive { it ->
```

Comment:
why use an explicit variable `it` so much?

Reply:
removed, I don't remember hot it end up like this.